### PR TITLE
coretasks: trying to improve docstrings

### PR DESCRIPTION
**Pull request by Exirel**
### Description

Working on #1565, for the `coretasks` core plugin. I wanted to do a quick first draft for @dgw to review.

I'm not really proud of these docstrings. On the other hand, I think we need a full chapter to explain how Sopel initialize the connection, deal with SASL, WHO/WHOX/NAMES, and stuffs. So it's not really my biggest concern.

My biggest concern is the lack of test for this plugin. I might add them here, maybe. We'll see.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

File: sopel/coretasks.py [hunk 1]
```diff
@@ -1,9 +1,17 @@
 # coding=utf-8
-"""Tasks that allow the bot to run, but aren't user-facing functionality
+"""Core Sopel plugin that listen to basic IRC messages.
```
Comment (dgw) [inline sopel/coretasks.py:2]:
```suggestion
"""Core Sopel plugin that listens to basic IRC messages.
```

But actually, I'd go further:

```suggestion
"""Core Sopel plugin that handles IRC protocol functions.
```

File: sopel/coretasks.py [hunk 2]
```diff
@@ -1,9 +1,17 @@
...
+
+This plugin allows the bot to run without user-facing functionality:
+
+* it handles client capability negotiation,
+* it handles client auth (both nick auth and server auth),
+* it handles connection registration (RPL_WELCOME, RPL_LUSERCLIENT), dealing
+  with error cases such as nick already in use
+* it tracks known channels & users (join, quit, nick change and other events).
```
Comment (dgw) [inline sopel/coretasks.py:10]:
```suggestion
* it tracks known channels & users (join, quit, nick change and other events)
* it manages blocked (ignored) users
```

Until/unless the `.blocks` stuff gets moved out of this file, it should get mentioned here.

File: sopel/coretasks.py [hunk 3]
```diff
@@ -1,9 +1,17 @@
 # coding=utf-8
-"""Tasks that allow the bot to run, but aren't user-facing functionality
+"""Core Sopel plugin that listen to basic IRC messages.
+
+This plugin allows the bot to run without user-facing functionality:
+
+* it handles client capability negotiation,
+* it handles client auth (both nick auth and server auth),
```
Comment (dgw) [inline sopel/coretasks.py:7]:
```suggestion
* it handles client capability negotiation
* it handles client auth (both nick auth and server auth)
```

I've never been a fan of this style of list, and the remaining items didn't really carry through the pattern anyway. x)

File: sopel/coretasks.py [hunk 4]
```diff
@@ -41,6 +49,11 @@
 
 
 def setup(bot):
+    """Setup the coretasks plugin.
```
Comment (dgw) [inline sopel/coretasks.py:52]: Per [setup (v.)](https://en.wiktionary.org/wiki/setup#Verb):

```suggestion
    """Set up the coretasks plugin.
```

File: sopel/coretasks.py [hunk 5]
```diff
@@ -59,6 +72,7 @@ def processing_job(bot):
 
 
 def shutdown(bot):
+    """Clean-up memory after the plugin."""
```
Comment (dgw) [inline sopel/coretasks.py:75]:
```suggestion
    """Clean up coretasks-related values in the bot's memory."""
```

File: sopel/coretasks.py [hunk 6]
```diff
@@ -84,7 +98,25 @@ def _join_event_processing(bot):
 
 
 def auth_after_register(bot):
-    """Do NickServ/AuthServ auth"""
+    """Do NickServ/AuthServ auth
```
Comment (dgw) [inline sopel/coretasks.py:101]:
```suggestion
    """Do NickServ/AuthServ auth.
```

File: sopel/coretasks.py [hunk 7]
```diff
@@ -84,7 +98,25 @@ def _join_event_processing(bot):
...
+
+    :param bot: a connected Sopel instance
+    :type bot: :class:`sopel.bot.Sopel`
+
+    This function can be used, **after** the bot is connected, to handle one of
+    these auth methods:
+
+    * ``nickserv``: send a private message to the NickServ service
+    * ``authserv``: send a ``AUTHSERV`` command
+    * ``Q``: send a ``AUTH`` command
```
Comment (dgw) [inline sopel/coretasks.py:111]:
```suggestion
    * ``authserv``: send an ``AUTHSERV`` command
    * ``Q``: send an ``AUTH`` command
```

f-english 😁

File: sopel/coretasks.py [hunk 8]
```diff
@@ -120,15 +152,28 @@ def auth_after_register(bot):
 
 
 def _execute_perform(bot):
-    """Execute commands specified to perform on IRC server connect."""
+    """Execute commands specified to perform on IRC server connect.
+
+    This function executes the list of commands that can be found in the
+    ``core.commands_on_connect`` setting. It automatically replace any
```
Comment (dgw) [inline sopel/coretasks.py:158]:
```suggestion
    ``core.commands_on_connect`` setting. It automatically replaces any
```

File: sopel/coretasks.py [hunk 9]
```diff
@@ -120,15 +152,28 @@ def auth_after_register(bot):
...
+        LOGGER.info('No custom command to execute.')
+        return
+
+    LOGGER.info('Executing %d custom commands.', count)
+    for i, command in enumerate(commands, 1):
         command = command.replace('$nickname', bot.config.core.nick)
-        LOGGER.debug(command)
+        LOGGER.debug(
+            'Executing custom command [%d/%d]: %s', i, count, command)
         bot.write((command,))
```
Comment (dgw) [inline sopel/coretasks.py:177]: This mini-refactor belongs in its own commit.

File: sopel/coretasks.py [hunk 10]
```diff
@@ -137,6 +182,16 @@ def _execute_perform(bot):
 @plugin.priority('high')
 @plugin.unblockable
 def on_nickname_in_use(bot, trigger):
+    """Change the bot's nick when the current one is already in use.
+
+    This can be triggered when the bot disconnect then reconnect before the
```
Comment (dgw) [inline sopel/coretasks.py:187]:
```suggestion
    This can be triggered when the bot disconnects then reconnects before the
```

File: sopel/coretasks.py [hunk 11]
```diff
@@ -137,6 +182,16 @@ def _execute_perform(bot):
...
+    server can notice a client timeout. Other reasons include mischief,
+    trolling, and obviously, PEBKAC.
+
+    This will change the current nick by adding a trailing ``_``. If a the bot
+    notices that a user with the configured nick disconnect (see ``QUIT`` event
+    handling), it will try to change it back.
```
Comment (dgw) [inline sopel/coretasks.py:193]:
```suggestion
    This will change the current nick by adding a trailing ``_``. If the bot
    sees that a user with its configured nick disconnects (see ``QUIT`` event
    handling), the bot will try to regain it.
```

File: sopel/coretasks.py [hunk 12]
```diff
@@ -151,7 +206,11 @@ def on_nickname_in_use(bot, trigger):
 @module.require_admin("This command requires admin privileges.")
 @module.commands('execute')
 def execute_perform(bot, trigger):
-    """Execute commands specified to perform on IRC server connect."""
+    """Execute commands specified to perform on IRC server connect.
+
+    Note that this allows a bot owner or admin to force the execution of
+    commands that are automatically performed when the bot connect.
```
Comment (dgw) [inline sopel/coretasks.py:212]:
```suggestion
    This allows a bot owner or admin to force the execution of commands
    that are automatically performed when the bot connects.
```

File: sopel/coretasks.py [hunk 13]
```diff
@@ -162,28 +221,41 @@ def execute_perform(bot, trigger):
 def startup(bot, trigger):
     """Do tasks related to connecting to the network.
 
-    001 RPL_WELCOME is from RFC2812 and is the first message that is sent after
-    the connection has been registered on the network.
+    ``001 RPL_WELCOME`` is from RFC2812 and is the first message that is sent
```
Comment (dgw) [inline sopel/coretasks.py:224]: I smell a future project to convert RFC references throughout the docs to Sphinx extlinks. Like:

```rst
:rfc:`2812`
```

File: sopel/coretasks.py [hunk 14]
```diff
@@ -162,28 +221,41 @@ def execute_perform(bot, trigger):
...
+    after the connection has been registered on the network.
 
-    251 RPL_LUSERCLIENT is a mandatory message that is sent after client
+    ``251 RPL_LUSERCLIENT`` is a mandatory message that is sent after client
     connects to the server in rfc1459. RFC2812 does not require it and all
```
Comment (dgw) [inline sopel/coretasks.py:228]:
```suggestion
    connects to the server in RFC1459. RFC2812 does not require it and all
```

File: sopel/coretasks.py [hunk 15]
```diff
@@ -162,28 +221,41 @@ def execute_perform(bot, trigger):
...
     networks might not send it. We support both.
 
+    If ``sopel.irc.AbstractBot.connection_registered`` is set this function
+    does nothing and returns immediatly. Otherwise, the flag is set and the
+    function proceed normally to:
```
Comment (dgw) [inline sopel/coretasks.py:233]:
```suggestion
    does nothing and returns immediately. Otherwise, the flag is set and the
    function proceeds normally to:
```

File: sopel/coretasks.py [hunk 16]
```diff
@@ -162,28 +221,41 @@ def execute_perform(bot, trigger):
 def startup(bot, trigger):
     """Do tasks related to connecting to the network.
 
-    001 RPL_WELCOME is from RFC2812 and is the first message that is sent after
-    the connection has been registered on the network.
+    ``001 RPL_WELCOME`` is from RFC2812 and is the first message that is sent
+    after the connection has been registered on the network.
 
-    251 RPL_LUSERCLIENT is a mandatory message that is sent after client
+    ``251 RPL_LUSERCLIENT`` is a mandatory message that is sent after client
```
Comment (dgw) [inline sopel/coretasks.py:227]:
```suggestion
    ``251 RPL_LUSERCLIENT`` is a mandatory message that is sent after the client
```

I want this, but I am far too lazy to rewrap the entire paragraph in a suggestion while sitting in a laundromat wearing my COVID-appropriate face mask. 😷 (Also I want to get through the rest of the PR.)

File: sopel/coretasks.py [hunk 17]
```diff
@@ -162,28 +221,41 @@ def execute_perform(bot, trigger):
...
     connects to the server in rfc1459. RFC2812 does not require it and all
     networks might not send it. We support both.
 
+    If ``sopel.irc.AbstractBot.connection_registered`` is set this function
+    does nothing and returns immediatly. Otherwise, the flag is set and the
+    function proceed normally to:
+
+    1. trigger auth method,
+    2. set bot's ``MODE`` (from ``core.modes``),
```
Comment (dgw) [inline sopel/coretasks.py:236]:
```suggestion
    1. trigger auth method
    2. set bot's ``MODE`` (from ``core.modes``)
```

As above, dislike this style and the rest of the list didn't commit to it anyway.

File: sopel/coretasks.py [hunk 18]
```diff
@@ -264,6 +338,14 @@ def parse_reply_myinfo(bot, trigger):
 @module.require_owner()
 @module.commands('useserviceauth')
 def enable_service_auth(bot, trigger):
+    """Set owner's account from an authenticated owner.
+
+    This command can be used to automatically configure ``core.owner_account``
+    when the owner is known, has a registered account, but the bot doesn't
```
Comment (dgw) [inline sopel/coretasks.py:344]:
```suggestion
    when the owner is known and has a registered account, but the bot doesn't
```

File: sopel/coretasks.py [hunk 19]
```diff
@@ -497,6 +583,7 @@ def track_part(bot, trigger):
 @module.thread(False)
 @module.unblockable
 def track_kick(bot, trigger):
+    """Track users getting kick from channels."""
```
Comment (dgw) [inline sopel/coretasks.py:586]:
```suggestion
    """Track users kicked from channels."""
```

File: sopel/coretasks.py [hunk 20]
```diff
@@ -825,6 +919,17 @@ def send_authenticate(bot, token):
 @module.event('AUTHENTICATE')
 @module.unblockable
 def auth_proceed(bot, trigger):
+    """Handle client initiated authenticate.
```
Comment (dgw) [inline sopel/coretasks.py:922]:
```suggestion
    """Handle client-initiated SASL auth.
```

AFAIK, SASL is the only use for `AUTHENTICATE` in IRC specs. Did I miss something?

File: sopel/coretasks.py [hunk 21]
```diff
@@ -162,28 +221,41 @@ def execute_perform(bot, trigger):
...
-    the connection has been registered on the network.
+    ``001 RPL_WELCOME`` is from RFC2812 and is the first message that is sent
+    after the connection has been registered on the network.
 
-    251 RPL_LUSERCLIENT is a mandatory message that is sent after client
+    ``251 RPL_LUSERCLIENT`` is a mandatory message that is sent after client
     connects to the server in rfc1459. RFC2812 does not require it and all
     networks might not send it. We support both.
 
+    If ``sopel.irc.AbstractBot.connection_registered`` is set this function
```
Comment (dgw) [inline sopel/coretasks.py:231]:
```suggestion
    If ``sopel.irc.AbstractBot.connection_registered`` is set, this function
```

For consistency with `auth_proceed()` docstring below.

File: sopel/coretasks.py [hunk 22]
```diff
@@ -309,7 +391,10 @@ def retry_join(bot, trigger):
 @module.thread(False)
 @module.unblockable
 def handle_names(bot, trigger):
-    """Handle NAMES response, happens when joining to channels."""
+    """Handle NAMES response, happens when joining to channels.
+
+    This function keeps track of user's privileges.
```
Comment (dgw) [inline sopel/coretasks.py:396]:
```suggestion
    """Handle NAMES responses.

    This function keeps track of users' privileges when Sopel joins channels.
```

+wrap I guess, if needed

File: sopel/coretasks.py [hunk 23]
```diff
@@ -825,6 +919,17 @@ def send_authenticate(bot, token):
...
+
+    If the chosen mechanism is client-first, the server sends an empty
+    response (``AUTHENTICATE +``). In that case, Sopel will handle SASL auth
+    that uses a token.
+
+    .. important::
+
+        If ``core.auth_method`` is set, then ``core.server_auth_method`` will
+        be ignored. If none is set, then this function does nothing.
+    """
```
Comment (dgw) [inline sopel/coretasks.py:932]:
```suggestion
    
    """
```

For consistency.

File: sopel/coretasks.py [hunk 24]
```diff
@@ -849,6 +954,13 @@ def _make_sasl_plain_token(account, password):
 @module.event(events.RPL_SASLSUCCESS)
 @module.unblockable
 def sasl_success(bot, trigger):
+    """End CAP request on successful SASL auth.
+
+    If SASL is configured, then the bot won't send ``CAP END`` once it gets
+    all the capability responses: it will wait for SASL auth result.
```
Comment (dgw) [inline sopel/coretasks.py:960]:
```suggestion
    all the capability responses; it will wait for SASL auth result.
```

File: sopel/coretasks.py [hunk 25]
```diff
@@ -985,6 +1097,7 @@ def blocks(bot, trigger):
 
 @module.event('ACCOUNT')
 def account_notify(bot, trigger):
+    """Track users's account."""
```
Comment (dgw) [inline sopel/coretasks.py:1100]:
```suggestion
    """Track users' accounts."""
```

Is this a :fr: thing? 😹

File: sopel/coretasks.py [hunk 26]
```diff
@@ -998,6 +1111,7 @@ def account_notify(bot, trigger):
 @module.priority('high')
 @module.unblockable
 def recv_whox(bot, trigger):
+    """Track ``WHO`` responses when ``WHOX`` is enabled"""
```
Comment (dgw) [inline sopel/coretasks.py:1114]:
```suggestion
    """Track ``WHO`` responses when ``WHOX`` is enabled."""
```

File: sopel/coretasks.py [hunk 27]
```diff
@@ -1052,6 +1166,7 @@ def _record_who(bot, channel, user, host, nick, account=None, away=None, modes=N
 @module.priority('high')
 @module.unblockable
 def recv_who(bot, trigger):
+    """Track ``WHO`` responses when ``WHOX`` is not enabled"""
```
Comment (dgw) [inline sopel/coretasks.py:1169]:
```suggestion
    """Track ``WHO`` responses when ``WHOX`` is not enabled."""
```

File: sopel/coretasks.py [hunk 28]
```diff
@@ -1084,6 +1201,7 @@ def track_notify(bot, trigger):
 @module.thread(False)
 @module.unblockable
 def track_topic(bot, trigger):
+    """Track channels' topic."""
```
Comment (dgw) [inline sopel/coretasks.py:1204]:
```suggestion
    """Track channels' topics."""
```

File: sopel/coretasks.py [hunk 29]
```diff
@@ -1071,6 +1187,7 @@ def end_who(bot, trigger):
 @module.thread(False)
 @module.unblockable
 def track_notify(bot, trigger):
+    """Track user going away."""
```
Comment (dgw) [inline sopel/coretasks.py:1190]:
```suggestion
    """Track users going away or coming back."""
```

At least I _assume_ that `AWAY` events are sent also for coming back, based on the actual code.

**Review by dgw — CHANGES_REQUESTED** [review]
OK, this ended up a bit scarier in terms of line-note count than I expected based on the diff size. It's pretty much all trivial stuff, though. I like!

File: sopel/coretasks.py (hunk 2 above)
Comment (Exirel) [reply sopel/coretasks.py:10]: I knew I forgot something.

File: sopel/coretasks.py (hunk 9 above)
Comment (Exirel) [reply sopel/coretasks.py:177]: Yes... x)

File: sopel/coretasks.py (hunk 11 above)
Comment (Exirel) [reply sopel/coretasks.py:193]: "if a the bot" why did I do that here?

**Comment by Exirel** [comment]
> It's pretty much all trivial stuff, though. I like!

Maybe you didn't notice but I tried hard to not forget about S everywhere.  Also, I think I more or less remembered that "plurals' something" and not "plurals's something".

File: sopel/coretasks.py (hunk 9 above)
Comment (Exirel) [reply sopel/coretasks.py:178]: Which I've done!

File: sopel/coretasks.py (hunk 13 above)
Comment (Exirel) [reply sopel/coretasks.py:225]: Perfect for yet another PR.

**Comment by Exirel** [comment]
Fixed conflict and repushed.

**Comment by dgw** [comment]
I'm going to trust that Exi's rebase didn't break anything. Did skim the current changes, but not in-depth. (Who wants to re-review things? Not _this_ dgw!)

**Review by dgw — APPROVED** [review]
